app/controllers/albums.js

```diff
@@ -11,3 +17,32 @@ exports.getAlbumPhotos = (req, res, next) =>
   listAlbumPhotos(req.params.id)
     .then(response => res.status(200).send(response.body))
     .catch(next);
+
+exports.buyAlbum = (req, res, next) =>
+  getAlbumData(req.params.id)
+    .then(albumData =>
+      findAlbum(albumData.id, req.user.id).then(existingAlbum => {
+        if (existingAlbum) {
+          throw existingAlbumError(
+            `Album not bought: user ${req.user.firstName} ${req.user.lastName} already had '${albumData.title}'.`
+          );
+        } else {
+          return createAlbum(req.user.id, albumData.id, albumData.title)
+            .then(album => {
+              logger.info(
+                `Album bought: user ${req.user.firstName} ${req.user.lastName} now has '${albumData.title}'`
+              );
+              res.status(201).send(album);
+            })
+            .catch(next);
+        }
+      })
+    )
+    .catch(err => {
+      logger.error(util.inspect(err));
+      if (err.internalCode === 'api_albums_error') {
```

Comment:
Mmmm In general, you throw errors in services and in the controller always have `.catch(next)`